BUG: Windows build regressions

IPX is currently throwing the following error (see [logs](https://dev.azure.com/scipy-org/SciPy/_build/results?buildId=10798&view=logs&jobId=0c790cac-d60e-5695-2cee-2e3f569cb85a&j=0c790cac-d60e-5695-2cee-2e3f569cb85a&t=3337ac87-8da2-5fc3-12b4-7ce358a5e1cb)):
```
 scipy\optimize\_highs\src\ipm\ipx\src\guess_basis.cc(130): error C2039: 'max': is not a member of 'std'
  C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\VC\Tools\MSVC\14.16.27023\include\chrono(19): note: see declaration of 'std'
  scipy\optimize\_highs\src\ipm\ipx\src\guess_basis.cc(130): error C3861: 'max': identifier not found
```

Most likely a regression due to #439

--- a/app/RunHighs.cpp
+++ b/app/RunHighs.cpp
@@ -198,7 +198,7 @@ HighsStatus callLpSolver(HighsOptions& use_options, const HighsLp& lp) {
 
   // Solve LP case.
   Highs highs(use_options);
-  const HighsOptions& options = highs.getHighsOptions();
+  //  const HighsOptions& options = highs.getHighsOptions();
 
   // // Load problem.
   highs.passModel(lp);

--- a/check/TestLpSolvers.cpp
+++ b/check/TestLpSolvers.cpp
@@ -194,7 +194,7 @@ void testSolversSetup(const std::string model,
     simplex_strategy_iteration_count[(
         int)SimplexStrategy::SIMPLEX_STRATEGY_PRIMAL] = 94;
     model_iteration_count.ipm = 18;
-    model_iteration_count.crossover = 4;
+    model_iteration_count.crossover = 18; // Pushes 4; Updates
   }
 }
 

--- a/check/TestLpSolvers.cpp
+++ b/check/TestLpSolvers.cpp
@@ -194,7 +194,7 @@ void testSolversSetup(const std::string model,
     simplex_strategy_iteration_count[(
         int)SimplexStrategy::SIMPLEX_STRATEGY_PRIMAL] = 94;
     model_iteration_count.ipm = 18;
-    model_iteration_count.crossover = 18; // Pushes 4; Updates
+    model_iteration_count.crossover = 18;  // Pushes 4; Updates
   }
 }
 

--- a/src/io/HMpsFF.cpp
+++ b/src/io/HMpsFF.cpp
@@ -207,7 +207,7 @@ HMpsFF::parsekey HMpsFF::parseDefault(std::ifstream& file) {
     HMpsFF::parsekey key = checkFirstWord(strline, s, e, word);
     if (key == HMpsFF::parsekey::NAME) {
       // Save name of the MPS file
-      if (e < strline.length()) {
+      if (e < (int)strline.length()) {
         mpsName = first_word(strline, e);
       }
       return HMpsFF::parsekey::NONE;

--- a/src/ipm/IpxWrapper.h
+++ b/src/ipm/IpxWrapper.h
@@ -596,8 +596,8 @@ HighsStatus solveLpIpx(const HighsOptions& options, HighsTimer& timer,
   ipx::Info ipx_info = lps.GetInfo();
   iteration_counts.ipm += (int)ipx_info.iter;
 
-  iteration_counts.crossover += (int)ipx_info.updates_crossover;
-  // iteration_counts.crossover += (int)ipx_info.pushes_crossover;
+  // iteration_counts.crossover += (int)ipx_info.updates_crossover;
+  iteration_counts.crossover += (int)ipx_info.pushes_crossover;
 
   // If not solved...
   if (solve_status != IPX_STATUS_solved) {

--- a/src/ipm/IpxWrapper.h
+++ b/src/ipm/IpxWrapper.h
@@ -584,7 +584,7 @@ HighsStatus solveLpIpx(const HighsOptions& options, HighsTimer& timer,
                   "IPX model has %d rows, %d columns and %d nonzeros",
                   (int)num_row, (int)num_col, (int)Ap[num_col]);
 
-  ipx::Int load_status =
+  //  ipx::Int load_status =
       lps.LoadModel(num_col, &objective[0], &col_lb[0], &col_ub[0], num_row,
                     &Ap[0], &Ai[0], &Av[0], &rhs[0], &constraint_type[0]);
 

--- a/src/ipm/IpxWrapper.h
+++ b/src/ipm/IpxWrapper.h
@@ -585,8 +585,8 @@ HighsStatus solveLpIpx(const HighsOptions& options, HighsTimer& timer,
                   (int)num_row, (int)num_col, (int)Ap[num_col]);
 
   //  ipx::Int load_status =
-      lps.LoadModel(num_col, &objective[0], &col_lb[0], &col_ub[0], num_row,
-                    &Ap[0], &Ai[0], &Av[0], &rhs[0], &constraint_type[0]);
+  lps.LoadModel(num_col, &objective[0], &col_lb[0], &col_ub[0], num_row, &Ap[0],
+                &Ai[0], &Av[0], &rhs[0], &constraint_type[0]);
 
   // todo: handle load error
   ipx::Int solve_status = lps.Solve();

--- a/src/ipm/ipx/include/ipx_info.h
+++ b/src/ipm/ipx/include/ipx_info.h
@@ -56,6 +56,7 @@ struct ipx_info {
     ipxint updates_start;       /* # basis updates for starting basis */
     ipxint updates_ipm;         /* # basis updates in IPM */
     ipxint updates_crossover;   /* # basis updates in crossover */
+    ipxint pushes_crossover;    /* # Primal/Dual pushes in crossover */
 
     /* major computation times */
     double time_total;          /* total runtime (wallclock) */

--- a/src/ipm/ipx/src/basiclu_kernel.cc
+++ b/src/ipm/ipx/src/basiclu_kernel.cc
@@ -67,6 +67,7 @@ void BasicLuKernel::_Factorize(Int dim, const Int* Bbegin, const Int* Bend,
     Int num_dropped = RemoveDiagonal(*L, nullptr);
     assert(num_dropped == dim);
     assert(L->entries() == lu.obj.xstore[BASICLU_LNZ]);
+    (void)(num_dropped);
 }
 
 }  // namespace ipx

--- a/src/ipm/ipx/src/basiclu_wrapper.cc
+++ b/src/ipm/ipx/src/basiclu_wrapper.cc
@@ -111,6 +111,7 @@ void BasicLu::_GetFactors(SparseMatrix* L, SparseMatrix* U, Int* rowperm,
         // Remove unit diagonal from L.
         Int num_dropped = RemoveDiagonal(*L, nullptr);
         assert(num_dropped == dim);
+        (void)(num_dropped);
     }
     if (dependent_cols) {
         // Dependent columns are at the end of the BASICLU pivot sequence.

--- a/src/ipm/ipx/src/basis.cc
+++ b/src/ipm/ipx/src/basis.cc
@@ -76,6 +76,7 @@ void Basis::SetToSlackBasis() {
     Int err = Factorize();
     // factorization of slack basis cannot fail other than out of memory
     assert(err == 0);
+    (void)(err);
 }
 
 Int Basis::Load(const int* basic_status) {
@@ -251,7 +252,7 @@ void Basis::TableauRow(Int jb, IndexedVector& btran, IndexedVector& row,
             Int end = AIt.end(i);
             for (Int p = begin; p < end; p++) {
                 Int j = Ati[p];
-                if (map2basis_[j] == -1 || map2basis_[j] == -2 && !ignore_fixed)
+                if (map2basis_[j] == -1 || (map2basis_[j] == -2 && !ignore_fixed))
                 {
                     map2basis_[j] -= 2; // mark column
                     row_pattern[nz++] = j;
@@ -270,7 +271,7 @@ void Basis::TableauRow(Int jb, IndexedVector& btran, IndexedVector& row,
         const double* Ax = AI.values();
         for (Int j = 0; j < n+m; j++) {
             double result = 0.0;
-            if (map2basis_[j] == -1 || map2basis_[j] == -2 && !ignore_fixed) {
+            if (map2basis_[j] == -1 || (map2basis_[j] == -2 && !ignore_fixed)) {
                 Int begin = AI.begin(j);
                 Int end = AI.end(j);
                 for (Int p = begin; p < end; p++)
@@ -347,8 +348,8 @@ void Basis::ComputeBasicSolution(Vector& x, Vector& y, Vector& z) const {
 }
 
 void Basis::ConstructBasisFromWeights(const double* colscale, Info* info) {
-    const Int m = model_.rows();
-    const Int n = model_.cols();
+    // const Int m = model_.rows();
+    // const Int n = model_.cols();
     assert(colscale);
     info->errflag = 0;
     info->dependent_rows = 0;
@@ -508,15 +509,15 @@ void Basis::CrashBasis(const double* colweights) {
     // rank revealing factorization, but it detects many dependencies in
     // practice.
     std::vector<Int> cols_guessed = GuessBasis(control_, model_, colweights);
-    assert(cols_guessed.size() <= m);
-    assert(cols_guessed.size() == m); // at the moment
+    assert((int)cols_guessed.size() <= m);
+    assert((int)cols_guessed.size() == m); // at the moment
 
     // Initialize the Basis object and factorize the (partial) basis. If
     // basis_[p] is negative, the p-th column of the basis matrix is zero,
     // and a slack column will be inserted by CrashFacorize().
     std::fill(basis_.begin(), basis_.end(), -1);
     std::fill(map2basis_.begin(), map2basis_.end(), -1);
-    for (Int k = 0; k < cols_guessed.size(); k++) {
+    for (Int k = 0; k < (Int) cols_guessed.size(); k++) {
         basis_[k] = cols_guessed[k];
         assert(map2basis_[basis_[k]] == -1); // must not have duplicates
         map2basis_[basis_[k]] = k;
@@ -526,6 +527,7 @@ void Basis::CrashBasis(const double* colweights) {
     control_.Debug()
         << Textline("Number of columns dropped from guessed basis:")
         << num_dropped << '\n';
+    (void)(m);
 }
 
 // Rook search for a large entry in inverse(B). If successful, returns [p,i,x]

--- a/src/ipm/ipx/src/crossover.cc
+++ b/src/ipm/ipx/src/crossover.cc
@@ -29,7 +29,7 @@ void Crossover::PushAll(Basis* basis, Vector& x, Vector& y, Vector& z,
 
     // Run dual push phase.
     std::vector<Int> dual_superbasics;
-    for (Int p = 0; p < perm.size(); p++) {
+    for (Int p = 0; p < (Int) perm.size(); p++) {
         Int j = perm[p];
         if (basis->IsBasic(j) && z[j] != 0.0)
             dual_superbasics.push_back(j);
@@ -113,13 +113,13 @@ void Crossover::PushPrimal(Basis* basis, Vector& x,
 
     control_.ResetPrintInterval();
     Int next = 0;
-    while (next < variables.size()) {
+    while (next < (Int) variables.size()) {
         if ((info->errflag = control_.InterruptCheck()) != 0)
             break;
 
         const Int jn = variables[next];
         if (x[jn] == lb[jn] || x[jn] == ub[jn] ||
-            x[jn] == 0.0 && std::isinf(lb[jn]) && std::isinf(ub[jn])) {
+            (x[jn] == 0.0 && std::isinf(lb[jn]) && std::isinf(ub[jn]))) {
             // nothing to do
             next++;
             continue;
@@ -242,15 +242,15 @@ void Crossover::PushDual(Basis* basis, Vector& y, Vector& z,
             throw std::logic_error("invalid variable in Crossover::PushDual");
     }
     for (Int j = 0; j < n+m; j++) {
-        if ((sign_restrict[j] & 1) && z[j] < 0.0 ||
-            (sign_restrict[j] & 2) && z[j] > 0.0)
+        if (((sign_restrict[j] & 1) && z[j] < 0.0) ||
+            ((sign_restrict[j] & 2) && z[j] > 0.0))
             throw std::logic_error(
                 "sign condition violated in Crossover::PushDual");
     }
 
     control_.ResetPrintInterval();
     Int next = 0;
-    while (next < variables.size()) {
+    while (next < (Int) variables.size()) {
         if ((info->errflag = control_.InterruptCheck()) != 0)
             break;
 
@@ -346,7 +346,7 @@ void Crossover::PushDual(Basis* basis, Vector& y, Vector& z,
     const Vector& ub = model.ub();
 
     std::vector<int> sign_restrict(n+m);
-    for (Int j = 0; j < sign_restrict.size(); j++) {
+    for (Int j = 0; j < (Int) sign_restrict.size(); j++) {
         if (x[j] != ub[j]) sign_restrict[j] |= 1;
         if (x[j] != lb[j]) sign_restrict[j] |= 2;
     }

--- a/src/ipm/ipx/src/diagonal_precond.cc
+++ b/src/ipm/ipx/src/diagonal_precond.cc
@@ -55,8 +55,8 @@ void DiagonalPrecond::_Apply(const Vector& rhs, Vector& lhs,
     Timer timer;
 
     assert(factorized_);
-    assert(lhs.size() == m);
-    assert(rhs.size() == m);
+    assert((int)lhs.size() == m);
+    assert((int)rhs.size() == m);
 
     for (Int i = 0; i < m; i++) {
         lhs[i] = rhs[i] / diagonal_[i];

--- a/src/ipm/ipx/src/forrest_tomlin.cc
+++ b/src/ipm/ipx/src/forrest_tomlin.cc
@@ -254,7 +254,7 @@ void ForrestTomlin::SolvePermuted(Vector& lhs, char trans) {
     assert(U_.cols() == dim_+num_updates);
 
     // Require num_updates elements workspace at end of lhs.
-    assert(lhs.size() >= dim_+num_updates);
+    assert((int)lhs.size() >= dim_+num_updates);
 
     if (trans == 't' || trans == 'T') {
         // Move replaced entries to the end of the pivot sequence and zero out
@@ -318,7 +318,7 @@ void ForrestTomlin::ComputeSpike(Int nb, const Int* bi, const double* bx) {
 
     // Store spike in U. Indices are sorted, which is required for the sparse
     // dot product in Update().
-    Int nz = 0;
+    // Int nz = 0;
     U_.clear_queue();
     for (Int p = 0; p < dim_+num_updates; p++) {
         if (work_[p] != 0.0)
@@ -350,7 +350,7 @@ void ForrestTomlin::ComputeEta(Int j) {
 
     // Queue eta at end of R. Indices are sorted, which is required for the
     // sparse dot product in Update().
-    Int nz = 0;
+    // Int nz = 0;
     R_.clear_queue();
     double pivot = work_[pos];
     for (Int i = pos+1; i < dim_+num_updates; i++) {

--- a/src/ipm/ipx/src/guess_basis.cc
+++ b/src/ipm/ipx/src/guess_basis.cc
@@ -1,5 +1,6 @@
 // Copyright (c) 2018 ERGO-Code. See license.txt for license.
 
+#include <algorithm>
 #include "guess_basis.h"
 #include <cassert>
 #include <cmath>
@@ -186,7 +187,7 @@ static void ProcessRemaining(const Control& control, const Model& model,
         } else {
             num_failed++;
         }
-        if (num_failed >= 10*(m-basis->size()))
+        if (num_failed >= 10*(m - (Int) basis->size()))
             break;
     }
     for (Int i = 0; i < m; i++) {
@@ -227,7 +228,7 @@ std::vector<Int> GuessBasis(const Control& control, const Model& model,
         if (rownumber[i] < 0)
             basis.push_back(n+i);
     }
-    assert(basis.size() == m);
+    assert((int)basis.size() == m);
     return basis;
 }
 

--- a/src/ipm/ipx/src/info.cc
+++ b/src/ipm/ipx/src/info.cc
@@ -64,6 +64,7 @@ std::ostream& operator<<(std::ostream& os, const Info& info) {
     dump(os, "updates_start", info.updates_start);
     dump(os, "updates_ipm", info.updates_ipm);
     dump(os, "updates_crossover", info.updates_crossover);
+    dump(os, "pushes_crossover", info.pushes_crossover);
 
     dump(os, "time_total", fix2(info.time_total));
     dump(os, "time_ipm1", fix2(info.time_ipm1));

--- a/src/ipm/ipx/src/ipm.cc
+++ b/src/ipm/ipx/src/ipm.cc
@@ -358,7 +358,7 @@ void IPM::StepSizes(const Step& step) {
     const Vector& dxu = step.xu;
     const Vector& dzl = step.zl;
     const Vector& dzu = step.zu;
-    const double mu = iterate_->mu();
+    //    const double mu = iterate_->mu();
     const double gammaf = 0.9;
     const double gammaa = 1.0 / (1.0-gammaf);
 

--- a/src/ipm/ipx/src/ipm.cc
+++ b/src/ipm/ipx/src/ipm.cc
@@ -16,6 +16,7 @@ struct IPM::Step {
     Step& operator+=(const Step& rhs) {
         x += rhs.x; xl += rhs.xl; xu += rhs.xu;
         y += rhs.y; zl += rhs.zl; zu += rhs.zu;
+        return *this;
     }
 };
 

--- a/src/ipm/ipx/src/iterate.cc
+++ b/src/ipm/ipx/src/iterate.cc
@@ -1,5 +1,6 @@
 // Copyright (c) 2018 ERGO-Code. See license.txt for license.
 
+#include <algorithm>
 #include "iterate.h"
 #include <cassert>
 #include <cmath>
@@ -64,12 +65,12 @@ void Iterate::Initialize(const Vector& x, const Vector& xl, const Vector& xu,
     const Int n = model_.cols();
     const Vector& lb = model_.lb();
     const Vector& ub = model_.ub();
-    assert(x.size() == n+m);
-    assert(xl.size() == n+m);
-    assert(xu.size() == n+m);
-    assert(y.size() == m);
-    assert(zl.size() == n+m);
-    assert(zu.size() == n+m);
+    assert((int)x.size() == n+m);
+    assert((int)xl.size() == n+m);
+    assert((int)xu.size() == n+m);
+    assert((int)y.size() == m);
+    assert((int)zl.size() == n+m);
+    assert((int)zu.size() == n+m);
     x_ = x; xl_ = xl; xu_ = xu; y_ = y; zl_ = zl; zu_ = zu;
 
     // Set variable statuses.
@@ -323,9 +324,9 @@ void Iterate::DropToComplementarity(Vector& x, Vector& y, Vector& z) const {
     const Vector& ub = model_.ub();
     assert(postprocessed_);
 
-    assert(x.size() == n+m);
-    assert(y.size() == m);
-    assert(z.size() == n+m);
+    assert((int)x.size() == n+m);
+    assert((int)y.size() == m);
+    assert((int)z.size() == n+m);
     y = y_;
     for (Int j = 0; j < n+m; j++) {
         const double xlj = xl_[j];
@@ -429,6 +430,8 @@ void Iterate::ResidualsFromDropping(double* pres, double* dres) const {
                     zdrop = zl_[j]-zu_[j]; // inactive
             }
             break;
+        default:
+            break;
         }
         double amax = 0.0;
         for (Int p = AI.begin(j); p < AI.begin(j+1); p++)
@@ -517,6 +520,8 @@ void Iterate::assert_consistency() {
             assert(false);
         }
     }
+    (void)(lb);
+    (void)(ub);
 }
  
 void Iterate::Evaluate() const {

--- a/src/ipm/ipx/src/lp_solver.cc
+++ b/src/ipm/ipx/src/lp_solver.cc
@@ -90,7 +90,7 @@ Int LpSolver::Solve() {
         }
         PrintSummary();
     }
-    catch (std::bad_alloc) {
+    catch (const std::bad_alloc&) {
         control_.Log() << " out of memory\n";
         info_.status = IPX_STATUS_out_of_memory;
     }
@@ -485,6 +485,8 @@ void LpSolver::RunCrossover() {
             crossover.time_primal() + crossover.time_dual();
         info_.updates_crossover =
             crossover.primal_pivots() + crossover.dual_pivots();
+        info_.pushes_crossover =
+            crossover.primal_pushes() + crossover.dual_pushes();
         if (info_.status_crossover != IPX_STATUS_optimal) {
             // Crossover failed. Discard solution.
             x_crossover_.resize(0);
@@ -497,7 +499,7 @@ void LpSolver::RunCrossover() {
     // Recompute vertex solution and set basic statuses.
     basis_->ComputeBasicSolution(x_crossover_, y_crossover_, z_crossover_);
     basic_statuses_.resize(n+m);
-    for (Int j = 0; j < basic_statuses_.size(); j++) {
+    for (Int j = 0; j < (Int) basic_statuses_.size(); j++) {
         if (basis_->IsBasic(j)) {
             basic_statuses_[j] = IPX_basic;
         } else {

--- a/src/ipm/ipx/src/model.cc
+++ b/src/ipm/ipx/src/model.cc
@@ -168,37 +168,37 @@ Int Model::PresolveIPMStartingPoint(const double* x_user,
     }
     for (Int j = 0; j < num_var_; ++j) {
         if (!(xl_temp[j] >= 0.0) ||
-            scaled_lbuser_[j] == -INFINITY && xl_temp[j] != INFINITY ||
-            scaled_lbuser_[j] != -INFINITY && xl_temp[j] == INFINITY)
+            (scaled_lbuser_[j] == -INFINITY && xl_temp[j] != INFINITY) ||
+            (scaled_lbuser_[j] != -INFINITY && xl_temp[j] == INFINITY))
             return IPX_ERROR_invalid_vector;
     }
     for (Int j = 0; j < num_var_; ++j) {
         if (!(xu_temp[j] >= 0.0) ||
-            scaled_ubuser_[j] == INFINITY && xu_temp[j] != INFINITY ||
-            scaled_ubuser_[j] != INFINITY && xu_temp[j] == INFINITY)
+            (scaled_ubuser_[j] == INFINITY && xu_temp[j] != INFINITY) ||
+            (scaled_ubuser_[j] != INFINITY && xu_temp[j] == INFINITY))
             return IPX_ERROR_invalid_vector;
     }
     for (Int i = 0; i < num_constr_; ++i) {
         if (!std::isfinite(slack_temp[i]) ||
-            constr_type_[i] == '=' && !(slack_temp[i] == 0.0) ||
-            constr_type_[i] == '<' && !(slack_temp[i] >= 0.0) ||
-            constr_type_[i] == '>' && !(slack_temp[i] <= 0.0) )
+            (constr_type_[i] == '=' && !(slack_temp[i] == 0.0)) ||
+	    (constr_type_[i] == '<' && !(slack_temp[i] >= 0.0)) ||
+	    (constr_type_[i] == '>' && !(slack_temp[i] <= 0.0)) )
             return IPX_ERROR_invalid_vector;
     }
     for (Int i = 0; i < num_constr_; ++i) {
         if (!std::isfinite(y_temp[i]) ||
-            constr_type_[i] == '<' && !(y_temp[i] <= 0.0) ||
-            constr_type_[i] == '>' && !(y_temp[i] >= 0.0) )
+            (constr_type_[i] == '<' && !(y_temp[i] <= 0.0)) ||
+	    (constr_type_[i] == '>' && !(y_temp[i] >= 0.0)) )
             return IPX_ERROR_invalid_vector;
     }
     for (Int j = 0; j < num_var_; ++j) {
         if (!(zl_temp[j] >= 0.0 && zl_temp[j] < INFINITY) ||
-            scaled_lbuser_[j] == -INFINITY && zl_temp[j] != 0.0)
+            (scaled_lbuser_[j] == -INFINITY && zl_temp[j] != 0.0))
             return IPX_ERROR_invalid_vector;
     }
     for (Int j = 0; j < num_var_; ++j) {
         if (!(zu_temp[j] >= 0.0 && zu_temp[j] < INFINITY) ||
-            scaled_ubuser_[j] == INFINITY && zu_temp[j] != 0.0)
+            (scaled_ubuser_[j] == INFINITY && zu_temp[j] != 0.0))
             return IPX_ERROR_invalid_vector;
     }
 

--- a/src/ipm/ipx/src/model.cc
+++ b/src/ipm/ipx/src/model.cc
@@ -109,9 +109,9 @@ void Model::PresolveStartingPoint(const double* x_user,
                                   Vector& z_solver) const {
     const Int m = rows();
     const Int n = cols();
-    assert(x_solver.size() == n+m);
-    assert(y_solver.size() == m);
-    assert(z_solver.size() == n+m);
+    assert((int)x_solver.size() == n+m);
+    assert((int)y_solver.size() == m);
+    assert((int)z_solver.size() == n+m);
 
     Vector x_temp(num_var_);
     Vector slack_temp(num_constr_);
@@ -128,6 +128,8 @@ void Model::PresolveStartingPoint(const double* x_user,
     ScalePoint(x_temp, slack_temp, y_temp, z_temp);
     DualizeBasicSolution(x_temp, slack_temp, y_temp, z_temp,
                          x_solver, y_solver, z_solver);
+    (void)(n);
+    (void)(m);
 }
 
 Int Model::PresolveIPMStartingPoint(const double* x_user,
@@ -221,12 +223,12 @@ void Model::PostsolveInteriorSolution(const Vector& x_solver,
                                       double* zl_user, double* zu_user) const {
     const Int m = rows();
     const Int n = cols();
-    assert(x_solver.size() == n+m);
-    assert(xl_solver.size() == n+m);
-    assert(xu_solver.size() == n+m);
-    assert(y_solver.size() == m);
-    assert(zl_solver.size() == n+m);
-    assert(zu_solver.size() == n+m);
+    assert((int)x_solver.size() == n+m);
+    assert((int)xl_solver.size() == n+m);
+    assert((int)xu_solver.size() == n+m);
+    assert((int)y_solver.size() == m);
+    assert((int)zl_solver.size() == n+m);
+    assert((int)zu_solver.size() == n+m);
 
     Vector x_temp(num_var_);
     Vector xl_temp(num_var_);
@@ -254,6 +256,8 @@ void Model::PostsolveInteriorSolution(const Vector& x_solver,
         std::copy(std::begin(zl_temp), std::end(zl_temp), zl_user);
     if (zu_user)
         std::copy(std::begin(zu_temp), std::end(zu_temp), zu_user);
+    (void)(n);
+    (void)(m);
 }
 
 void Model::EvaluateInteriorSolution(const Vector& x_solver,
@@ -265,12 +269,12 @@ void Model::EvaluateInteriorSolution(const Vector& x_solver,
                                      Info* info) const {
     const Int m = rows();
     const Int n = cols();
-    assert(x_solver.size() == n+m);
-    assert(xl_solver.size() == n+m);
-    assert(xu_solver.size() == n+m);
-    assert(y_solver.size() == m);
-    assert(zl_solver.size() == n+m);
-    assert(zu_solver.size() == n+m);
+    assert((int)x_solver.size() == n+m);
+    assert((int)xl_solver.size() == n+m);
+    assert((int)xu_solver.size() == n+m);
+    assert((int)y_solver.size() == m);
+    assert((int)zl_solver.size() == n+m);
+    assert((int)zu_solver.size() == n+m);
 
     // Build solution to scaled user model.
     Vector x(num_var_);
@@ -303,8 +307,8 @@ void Model::EvaluateInteriorSolution(const Vector& x_solver,
     }
     // rb = rhs-slack-A*x
     Vector rb = scaled_rhs_ - slack;
-    assert(scaled_rhs_.size() == num_constr_);
-    assert(rb.size() == num_constr_);
+    assert((int)scaled_rhs_.size() == num_constr_);
+    assert((int)rb.size() == num_constr_);
     MultiplyWithScaledMatrix(x, -1.0, rb, 'N');
     // rc = obj-zl+zu-A'y
     Vector rc = scaled_obj_ - zl + zu;
@@ -352,6 +356,8 @@ void Model::EvaluateInteriorSolution(const Vector& x_solver,
     info->normx = Infnorm(x);
     info->normy = Infnorm(y);
     info->normz = std::max(Infnorm(zl), Infnorm(zu));
+    (void)(n);
+    (void)(m);
 }
 
 void Model::PostsolveBasicSolution(const Vector& x_solver,
@@ -362,10 +368,10 @@ void Model::PostsolveBasicSolution(const Vector& x_solver,
                                    double* y_user, double* z_user) const {
     const Int m = rows();
     const Int n = cols();
-    assert(x_solver.size() == n+m);
-    assert(y_solver.size() == m);
-    assert(z_solver.size() == n+m);
-    assert(basic_status_solver.size() == n+m);
+    assert((int)x_solver.size() == n+m);
+    assert((int)y_solver.size() == m);
+    assert((int)z_solver.size() == n+m);
+    assert((int)basic_status_solver.size() == n+m);
 
     Vector x_temp(num_var_);
     Vector slack_temp(num_constr_);
@@ -387,6 +393,8 @@ void Model::PostsolveBasicSolution(const Vector& x_solver,
         std::copy(std::begin(y_temp), std::end(y_temp), y_user);
     if (z_user)
         std::copy(std::begin(z_temp), std::end(z_temp), z_user);
+    (void)(n);
+    (void)(m);
 }
 
 void Model::EvaluateBasicSolution(const Vector& x_solver,
@@ -396,10 +404,10 @@ void Model::EvaluateBasicSolution(const Vector& x_solver,
                                   Info* info) const {
     const Int m = rows();
     const Int n = cols();
-    assert(x_solver.size() == n+m);
-    assert(y_solver.size() == m);
-    assert(z_solver.size() == n+m);
-    assert(basic_status_solver.size() == n+m);
+    assert((int)x_solver.size() == n+m);
+    assert((int)y_solver.size() == m);
+    assert((int)z_solver.size() == n+m);
+    assert((int)basic_status_solver.size() == n+m);
 
     // Build basic solution to scaled user model.
     Vector x(num_var_);
@@ -450,13 +458,15 @@ void Model::EvaluateBasicSolution(const Vector& x_solver,
     info->primal_infeas = std::max(Infnorm(xinfeas), Infnorm(sinfeas));
     info->dual_infeas = std::max(Infnorm(zinfeas), Infnorm(yinfeas));
     info->objval = pobj;
+    (void)(n);
+    (void)(m);
 }
 
 void Model::PostsolveBasis(const std::vector<Int>& basic_status_solver,
                            Int* cbasis_user, Int* vbasis_user) const {
     const Int m = rows();
     const Int n = cols();
-    assert(basic_status_solver.size() == n+m);
+    assert((int)basic_status_solver.size() == n+m);
 
     std::vector<Int> cbasis_temp(num_constr_);
     std::vector<Int> vbasis_temp(num_var_);
@@ -466,6 +476,8 @@ void Model::PostsolveBasis(const std::vector<Int>& basic_status_solver,
         std::copy(std::begin(cbasis_temp), std::end(cbasis_temp), cbasis_user);
     if (vbasis_user)
         std::copy(std::begin(vbasis_temp), std::end(vbasis_temp), vbasis_user);
+    (void)(m);
+    (void)(n);
 }
 
 // Checks if the vectors are valid LP data vectors. Returns 0 if OK and a
@@ -585,13 +597,13 @@ void Model::ScaleModel(const Control& control) {
 
     // Apply scaling to vectors.
     if (colscale_.size() > 0) {
-        assert(colscale_.size() == num_var_);
+        assert((int)colscale_.size() == num_var_);
         scaled_obj_ *= colscale_;
         scaled_lbuser_ /= colscale_;
         scaled_ubuser_ /= colscale_;
     }
     if (rowscale_.size() > 0) {
-        assert(rowscale_.size() == num_constr_);
+        assert((int)rowscale_.size() == num_constr_);
         scaled_rhs_ *= rowscale_;
     }
 }
@@ -836,10 +848,10 @@ void Model::PrintCoefficientRange(const Control& control) const {
     double amax = 0.0;
     for (Int j = 0; j < A_.cols(); j++) {
         for (Int p = A_.begin(j); p < A_.end(j); p++) {
-            double x = A_.value(p);
+            double x = std::abs(A_.value(p));
             if (x != 0.0) {
-                amin = std::min(amin, std::abs(x));
-                amax = std::max(amax, std::abs(x));
+                amin = std::min(amin, x);
+                amax = std::max(amax, x);
             }
         }
     }
@@ -1052,13 +1064,13 @@ void Model::DualizeBasicSolution(const Vector& x_user,
 
     if (dualized_) {
         assert(num_var_ == m);
-        assert(num_constr_ + boxed_vars_.size() == n);
+        assert(num_constr_ + (int)boxed_vars_.size() == n);
 
         // Build dual solver variables from primal user variables.
         y_solver = -x_user;
         for (Int i = 0; i < num_constr_; i++)
             z_solver[i] = -slack_user[i];
-        for (Int k = 0; k < boxed_vars_.size(); k++) {
+        for (Int k = 0; k < (Int) boxed_vars_.size(); k++) {
             Int j = boxed_vars_[k];
             z_solver[num_constr_+k] = c(num_constr_+k) + y_solver[j];
         }
@@ -1068,7 +1080,7 @@ void Model::DualizeBasicSolution(const Vector& x_user,
         // Build primal solver variables from dual user variables.
         std::copy_n(std::begin(y_user), num_constr_, std::begin(x_solver));
         std::copy_n(std::begin(z_user), num_var_, std::begin(x_solver) + n);
-        for (Int k = 0; k < boxed_vars_.size(); k++) {
+        for (Int k = 0; k < (Int) boxed_vars_.size(); k++) {
             Int j = boxed_vars_[k];
             if (x_solver[n+j] < 0.0) {
                 // j is a boxed variable and z_user[j] < 0
@@ -1172,7 +1184,7 @@ void Model::DualizeBackInteriorSolution(const Vector& x_solver,
 
     if (dualized_) {
         assert(num_var_ == m);
-        assert(num_constr_ + boxed_vars_.size() == n);
+        assert(num_constr_ + (int)boxed_vars_.size() == n);
         x_user = -y_solver;
 
         // If the solution from the solver would be exact, we could copy the
@@ -1311,7 +1323,7 @@ void Model::DualizeBackBasicSolution(const Vector& x_solver,
 
     if (dualized_) {
         assert(num_var_ == m);
-        assert(num_constr_ + boxed_vars_.size() == n);
+        assert(num_constr_ + (int)boxed_vars_.size() == n);
         x_user = -y_solver;
         for (Int i = 0; i < num_constr_; i++)
             slack_user[i] = -z_solver[i];
@@ -1331,6 +1343,7 @@ void Model::DualizeBackBasicSolution(const Vector& x_solver,
         std::copy_n(std::begin(y_solver), num_constr_, std::begin(y_user));
         std::copy_n(std::begin(z_solver), num_var_, std::begin(z_user));
     }
+    (void)(m);
 }
 
 void Model::DualizeBackBasis(const std::vector<Int>& basic_status_solver,
@@ -1341,7 +1354,7 @@ void Model::DualizeBackBasis(const std::vector<Int>& basic_status_solver,
 
     if (dualized_) {
         assert(num_var_ == m);
-        assert(num_constr_ + boxed_vars_.size() == n);
+        assert(num_constr_ + (int)boxed_vars_.size() == n);
         for (Int i = 0; i < num_constr_; i++) {
             if (basic_status_solver[i] == IPX_basic)
                 cbasis_user[i] = IPX_nonbasic;
@@ -1378,6 +1391,7 @@ void Model::DualizeBackBasis(const std::vector<Int>& basic_status_solver,
         for (Int j = 0; j < num_var_; j++)
             vbasis_user[j] = basic_status_solver[j];
     }
+    (void)(m);
 }
 
 void Model::CorrectScaledBasicSolution(Vector& x, Vector& slack, Vector& y,
@@ -1403,8 +1417,8 @@ void Model::CorrectScaledBasicSolution(Vector& x, Vector& slack, Vector& y,
 void Model::MultiplyWithScaledMatrix(const Vector& rhs, double alpha,
                                      Vector& lhs, char trans) const {
     if (trans == 't' || trans == 'T') {
-        assert(rhs.size() == num_constr_);
-        assert(lhs.size() == num_var_);
+        assert((int)rhs.size() == num_constr_);
+        assert((int)lhs.size() == num_var_);
         if (dualized())
             for (Int i = 0; i < num_constr_; i++)
                 ScatterColumn(AI_, i, alpha*rhs[i], lhs);
@@ -1413,8 +1427,8 @@ void Model::MultiplyWithScaledMatrix(const Vector& rhs, double alpha,
                 lhs[j] += alpha * DotColumn(AI_, j, rhs);
     }
     else {
-        assert(rhs.size() == num_var_);
-        assert(lhs.size() == num_constr_);
+        assert((int)rhs.size() == num_var_);
+        assert((int)lhs.size() == num_constr_);
         if (dualized())
             for (Int i = 0; i < num_constr_; i++)
                 lhs[i] += alpha * DotColumn(AI_, i, rhs);
@@ -1430,7 +1444,7 @@ double PrimalInfeasibility(const Model& model, const Vector& x) {
     assert(x.size() == lb.size());
 
     double infeas = 0.0;
-    for (Int j = 0; j < x.size(); j++) {
+    for (Int j = 0; j < (Int) x.size(); j++) {
         infeas = std::max(infeas, lb[j]-x[j]);
         infeas = std::max(infeas, x[j]-ub[j]);
     }
@@ -1445,7 +1459,7 @@ double DualInfeasibility(const Model& model, const Vector& x,
     assert(z.size() == lb.size());
 
     double infeas = 0.0;
-    for (Int j = 0; j < x.size(); j++) {
+    for (Int j = 0; j < (Int) x.size(); j++) {
         if (x[j] > lb[j])
             infeas = std::max(infeas, z[j]);
         if (x[j] < ub[j])
@@ -1457,10 +1471,10 @@ double DualInfeasibility(const Model& model, const Vector& x,
 double PrimalResidual(const Model& model, const Vector& x) {
     const SparseMatrix& AIt = model.AIt();
     const Vector& b = model.b();
-    assert(x.size() == AIt.rows());
+    assert((int)x.size() == AIt.rows());
 
     double res = 0.0;
-    for (Int i = 0; i < b.size(); i++) {
+    for (Int i = 0; i < (Int) b.size(); i++) {
         double r = b[i] - DotColumn(AIt, i, x);
         res = std::max(res, std::abs(r));
     }
@@ -1470,11 +1484,11 @@ double PrimalResidual(const Model& model, const Vector& x) {
 double DualResidual(const Model& model, const Vector& y, const Vector& z) {
     const SparseMatrix& AI = model.AI();
     const Vector& c = model.c();
-    assert(y.size() == AI.rows());
-    assert(z.size() == AI.cols());
+    assert((int)y.size() == AI.rows());
+    assert((int)z.size() == AI.cols());
 
     double res = 0.0;
-    for (Int j = 0; j < c.size(); j++) {
+    for (Int j = 0; j < (Int) c.size(); j++) {
         double r = c[j] - z[j] - DotColumn(AI, j, y);
         res = std::max(res, std::abs(r));
     }

--- a/src/ipm/ipx/src/normal_matrix.cc
+++ b/src/ipm/ipx/src/normal_matrix.cc
@@ -57,8 +57,8 @@ void NormalMatrix::_Apply(const Vector& rhs, Vector& lhs,
     Timer timer;
 
     assert(prepared_);
-    assert(lhs.size() == m);
-    assert(rhs.size() == m);
+    assert((int)lhs.size() == m);
+    assert((int)rhs.size() == m);
 
     if (W_) {
         #if MATVECMETHOD == 1

--- a/src/ipm/ipx/src/sparse_matrix.cc
+++ b/src/ipm/ipx/src/sparse_matrix.cc
@@ -196,29 +196,31 @@ void MultiplyAdd(const SparseMatrix& A, const Vector& rhs, double alpha,
     const Int m = A.rows();
     const Int n = A.cols();
     if (trans == 't' || trans == 'T') {
-        assert(rhs.size() == m);
-        assert(lhs.size() == n);
+        assert((int)rhs.size() == m);
+        assert((int)lhs.size() == n);
         for (Int j = 0; j < n; j++)
             lhs[j] += alpha * DotColumn(A, j, rhs);
     } else {
-        assert(rhs.size() == n);
-        assert(lhs.size() == m);
+        assert((int)rhs.size() == n);
+        assert((int)lhs.size() == m);
         for (Int j = 0; j < n; j++)
             ScatterColumn(A, j, alpha*rhs[j], lhs);
     }
+    (void)(m);
 }
 
 void AddNormalProduct(const SparseMatrix& A, const double* D, const Vector& rhs,
                       Vector& lhs) {
     const Int m = A.rows();
     const Int n = A.cols();
-    assert(rhs.size() == m);
-    assert(lhs.size() == m);
+    assert((int)rhs.size() == m);
+    assert((int)lhs.size() == m);
     for (Int j = 0; j < n; j++) {
         double temp = DotColumn(A, j, rhs);
         if (D) temp *= D[j]*D[j];
         ScatterColumn(A, j, temp, lhs);
     }
+    (void)(m);
 }
 
 Int TriangularSolve(const SparseMatrix& A, Vector& x, char trans,

--- a/src/ipm/ipx/src/splitted_normal_matrix.cc
+++ b/src/ipm/ipx/src/splitted_normal_matrix.cc
@@ -47,7 +47,7 @@ void SplittedNormalMatrix::Prepare(const Basis& basis, const double* colscale) {
     PermuteRows(N_, rowperm_inv_);
 
     // Scale columns of N.
-    for (Int k = 0; k < nonbasic_vars.size(); k++) {
+    for (Int k = 0; k < (Int) nonbasic_vars.size(); k++) {
         Int j = nonbasic_vars[k];
         double d = colscale[j];
         assert(std::isfinite(d));

--- a/src/ipm/ipx/src/starting_basis.cc
+++ b/src/ipm/ipx/src/starting_basis.cc
@@ -124,6 +124,8 @@ static void PostprocessDependencies(Iterate* iterate, Basis* p_basis,
         assert(zl[n+i] == 0.0);
         assert(zu[n+i] == 0.0);
     }
+    (void)(zl);
+    (void)(zu);
 }
 
 void StartingBasis(Iterate* iterate, Basis* p_basis, Info* info) {

--- a/src/ipm/ipx/src/symbolic_invert.cc
+++ b/src/ipm/ipx/src/symbolic_invert.cc
@@ -39,6 +39,7 @@ static std::vector<Int> Matching(const Model& model,
                                           work.data(), work2.data(),
                                           work3.data());
             assert(matched);
+            (void)(matched);
         }
     }
     for (Int j : basis) {
@@ -48,6 +49,7 @@ static std::vector<Int> Matching(const Model& model,
                                           work.data(), work2.data(),
                                           work3.data());
             assert(matched);
+            (void)(matched);
         }
     }
     return jmatch;
@@ -131,7 +133,7 @@ void SymbolicInvert(const Model& model, const std::vector<Int>& basis,
                     Int* rowcounts, Int* colcounts) {
     const SparseMatrix& AI = model.AI();
     const Int m = AI.rows();
-    assert(basis.size() == m);
+    assert((int)basis.size() == m);
 
     // jmatch is a permutation of basis such that B = AI[:,jmatch] has a
     // zero-free diagonal.

--- a/src/ipm/ipx/src/utils.cc
+++ b/src/ipm/ipx/src/utils.cc
@@ -39,7 +39,7 @@ double Infnorm(const Vector& x) {
 double Dot(const Vector& x, const Vector& y) {
     assert(x.size() == y.size());
     double d = 0.0;
-    for (Int i = 0; i < x.size(); i++)
+    for (Int i = 0; i < (Int) x.size(); i++)
         d += x[i]*y[i];
     return d;
 }
@@ -47,7 +47,7 @@ double Dot(const Vector& x, const Vector& y) {
 Int FindMaxAbs(const Vector& x) {
     double xmax = 0.0;
     Int imax = 0;
-    for (Int i = 0; i < x.size(); i++) {
+    for (Int i = 0; i < (Int) x.size(); i++) {
         if (std::abs(x[i]) > xmax) {
             xmax = std::abs(x[i]);
             imax = i;

